tf_quant_finance/black_scholes/approximations/american_option_test.py

```diff
@@ -204,6 +205,156 @@ def test_option_prices_all_call_options(self):
+       [0.05, 0.84, 4.40, 11.55, 20.69, 20.00, 10.17, 3.49, 0.77, 0.11]),
+      (0.08, 0.4, 0.25,
+       [1.29, 3.82, 8.35, 14.80, 22.71, 20.53, 12.91, 7.42, 3.93, 1.93]),
+      (0.08, 0.2, 0.5,
+       [0.41, 2.18, 6.50, 13.42, 22.06, 20.00, 10.70, 4.70, 1.71, 0.52]),
+  )
+  def test_bs1993_prices_pos_carries(self, discount_rates, volatilities,
+                                     expiries, expected_prices):
+    """Tests Bjerksund Stensland 1993 prices for positive cost_of_carries."""
+    cost_of_carries = np.array([0.04] * 10)
+    forwards = np.array([80.0, 90.0, 100.0, 110.0, 120.0] * 2)
+    strikes = np.array([100.0] * 10)
+    is_call_options = [True] * 5 + [False] * 5
+    computed_prices = bjerksund_stensland(
+        volatilities=volatilities,
+        strikes=strikes,
+        expiries=expiries,
+        discount_rates=discount_rates,
+        cost_of_carries=cost_of_carries,
+        forwards=forwards,
+        is_call_options=is_call_options,
+        modified_boundary=False,
+        dtype=tf.float64)
+    with self.subTest(name='ExpectedPrices'):
+      self.assertAllClose(expected_prices, computed_prices,
+                          rtol=5e-3, atol=5e-3,
+                          msg='Failed: Bjerksund Stensland 1993 positive carries test.')
+  
+  @parameterized.parameters(
+      (0.08, 0.08, 0.2, 0.25,
+       [20.00, 10.01, 3.16, 0.65, 0.09]),
+      (0.12, 0.12, 0.2, 0.25,
+       [20.00, 10.00, 2.86, 0.54, 0.07]),
+      (0.08, 0.08, 0.4, 0.25,
+       [20.28, 12.48, 7.04, 3.66, 1.77]),
+      (0.08, 0.08, 0.2, 0.5,
+       [20.00, 10.24, 4.11, 1.37, 0.39]),
+  )
+  def test_bs1993_prices_carries_equal_rate(self, discount_rates, cost_of_carries,
+                                      volatilities, expiries, expected_prices):
+    """Tests Bjerksund Stensland 1993 prices for zero cost_of_carries."""
+    forwards = np.array([80.0, 90.0, 100.0, 110.0, 120.0])
+    strikes = np.array([100.0] * 5)
+    is_call_options = np.array([False] * 5)
+    computed_prices = bjerksund_stensland(
+        volatilities=volatilities,
+        strikes=strikes,
+        expiries=expiries,
+        discount_rates=discount_rates,
+        cost_of_carries=cost_of_carries,
+        forwards=forwards,
+        is_call_options=is_call_options,
+        modified_boundary=False,
+        dtype=tf.float64)
+    with self.subTest(name='ExpectedPrices'):
+      self.assertAllClose(expected_prices, computed_prices,
+                          rtol=5e-3, atol=5e-3,
+                          msg='Failed: Bjerksund Stensland 1993 zero carries test.')
+  
+  @parameterized.parameters(
+      (0.08, -0.04, 0.2, 3.0,
+       [2.30, 4.71, 8.44, 13.74, 20.85, 25.61, 20.04, 15.47, 11.78, 8.87]),
+      (0.08, 0.0, 0.2, 3.0,
+       [3.95, 7.20, 11.64, 17.24, 23.93, 22.12, 16.14, 11.64, 8.31, 5.89]),
+      #(0.08, 0.04, 0.4, 3.0,
+      # [6.88, 11.49, 17.21, 23.84, 31.16, 20.32, 13.43, 8.86, 5.83, 3.83]),      # Results are suspect, see PR #54
```

Comment:
Any reasons why this test fails? Seems like the issue comes from `cost_of_carries ` which is positive in this case

Reply:
All of the tests are from the [reference document [1]](https://core.ac.uk/download/pdf/30824897.pdf), including this one.  I mentioned in the original post above that I believe these values to be suspect.  Here is a comparison of the values in the reference document and the values output by the code, alongside the outputs from Barone-Adesi Whaley and Black Scholes Merton:

```
Values from reference [1]:  Calls:  [ 6.88 11.49 17.21 23.84 31.16]  Puts:  [20.32 13.43  8.86  5.83  3.83]
Bjerksund-Stensland 1993    Calls:  [16.75 22.25 28.31 34.84 41.77]  Puts:  [28.42 23.97 20.34 17.34 14.86]
Barone-Adesi Whaley         Calls:  [12.62 17.02 21.93 27.29 33.05]  Puts:  [33.68 28.94 25.02 21.73 18.96]
Black Scholes Merton        Calls:  [12.26 16.54 21.31 26.52 32.09]  Puts:  [27.99 24.41 21.31 18.65 16.35]
```

It should be noted that the difference between Table 4, Section 3 (first row above) and Table 2, Section 3 is time, the former being `T = 3` and the latter being `T = 0.25`.  So the values output by the code with the same inputs except time match when `T` is short, but they do not match when `T` is long.  But Table 4 is made up of other options with `T = 3` and those values match.

This, along with the values in the reference document being rather low compared to other option pricing models, led me to my conclusion that the values in the reference document for this one section to be suspect - but I am open to any alternate theories.

Reply:
Ok, I see now. I have found a mistake in the paper. Volatility should be `0.2` in the paper instead of `0.4`. Fixing that leads to a correct result. I will amend that during the merge so no worries. One question though, it seems that the paper specifies spots `S` rather than forwards in the tables. Am I missing something?

Reply:
You are right, it should be `spots` and not `forwards`.  I originally used Haug (2007) and then switched halfway to the reference paper, as that is the original source, and got a bit mixed up.

I updated the the code, including the tests.  I also updated the volatility in the suspect test to 20% and it does indeed pass now.  I pushed the changes with those fixes.